ocaml/typing/ctype.ml

```diff
@@ -3375,6 +3377,18 @@ let unify3_var env jkind1 t1' t2 t2' =
         record_equation t1' t2';
       end
 
+(* This is used to check whether we should add a gadt equation refining a
+   Tconstr's jkind during pattern unification. *)
+let constr_jkind_refinable env t jkind =
+  let snap = Btype.snapshot () in
+  let refinable =
+    match unification_jkind_check env t jkind with
```

Comment:
Using `unification_jkind_check` here seems like overkill. That uses the `lmode` to determine whether to perform the check or delay/skip it. But I think we will always want to perform the checks in a GADT match... delaying wouldn't even really make sense here.

And we also don't need the full power of `constrain_type_jkind`, which deals specially with type variables. `check_type_jkind` should be sufficient, given that the type will always be a `Tconstr`.

So I don't think this is wrong, but it's using a bigger hammer than necessary... and would be wrong if the big hammer were necessary (that is, this would do the wrong thing if either `lmode` weren't `Perform_checks` or if the type is a variable).

Maybe fix up in a future patch?

Reply:
I considered trying to create a simplified version of some of the functions involved here, but judged that the clarity that comes from uniformly using the same function for jkind checks in unification was preferable.

Reply:
In particular I think it's important for correctness that the check here matches the check in unify3_var.  As you note, that one is dealing with more than tconstrs, but this solution seemed the cleanest way to achieve that.

Similarly, the backtracking here is unnecessary work, at least for now, considering these constraints.  But Leo pointed out that it made this more obviously correct.